### PR TITLE
Upgrade kernel compiler to GCC 7.

### PR DESCRIPTION
**Pull request by clefru**
###### Motivation for this change
Kernel spectre v2 hardening, see
https://github.com/NixOS/nixpkgs/issues/34383

###### Things done
* Compiled and booted into.
* Check that `modinfo zfs` shows `vermagic: 4.14.15 SMP mod_unload retpoline`.
* Check the output of `grep . /sys/devices/system/cpu/vulnerabilities/*`:
```
/sys/devices/system/cpu/vulnerabilities/meltdown:Not affected
/sys/devices/system/cpu/vulnerabilities/spectre_v1:Vulnerable
/sys/devices/system/cpu/vulnerabilities/spectre_v2:Mitigation: Full AMD retpoline
```

This is only PR is only for discussion. I found the use of stdenvKernelCc a bit ugly, but couldn't find a more idiomatic expression. 

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -13158,7 +13160,9 @@ with pkgs;
 
   # A function to build a manually-configured kernel
   linuxManualConfig = pkgs.buildLinux;
-  buildLinux = makeOverridable (callPackage ../os-specific/linux/kernel/manual-config.nix {});
+  buildLinux = makeOverridable (callPackage ../os-specific/linux/kernel/manual-config.nix {
+    buildPackages = buildPackages // { stdenv = overrideCC buildPackages.stdenv gcc7; };
```
Comment (Mic92) [inline pkgs/top-level/all-packages.nix]: Please add a comment here, why gcc7 is required.
This makes it easier to remove this line later, when gcc7 becomes default or is superseded by newer versions.

**Comment by fpletz** [comment]
I'm still hoping we can make gcc7 the default really soon so we don't have to add this hack. But if we can't make it for 18.03, this will come in handy.

The other question is if we should backport this to 17.09. Then pushing this hack and gcc 7.3.0 to 17.09 is probably the easiest thing to do.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -12970,10 +12970,12 @@ with pkgs;
      for a specific kernel.  This function can then be called for
      whatever kernel you're using. */
 
-  linuxPackagesFor = kernel: lib.makeExtensible (self: with self; {
+  linuxPackagesFor = kernel: let stdenvKernelCc = overrideCC stdenv kernel.buildCc;
```
Comment (clefru) [inline pkgs/top-level/all-packages.nix]: I haven't found a better way to express that other than with introducing a "let" here.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (clefru) [reply pkgs/top-level/all-packages.nix]: Added a comment.

**Comment by Mic92** [comment]
@fpletz if so, this should start soon. I expect breakage again with `gcc7`.

**Comment by clefru** [comment]
I ran the kernel-lts, kernel-latest and installer.simple tests. All succeed with a gcc7-compiled retpoline kernel. So, functionality-wise this PR can go in. If somebody has style comments, please add them.

Should we switch to gcc7 in general, only the last commit needs to be rolled-back. Using the kernel compiler for kernel module compilation should stay, as this is a sane thing to do.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (aszlig) [reply pkgs/top-level/all-packages.nix]: What about just `stdenv = overrideCC pkgs.stdenv kernel.buildCc;`?

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -12974,6 +12974,7 @@ with pkgs;
     callPackage = newScope self;
 
     inherit kernel;
+    stdenv = overrideCC pkgs.stdenv kernel.buildCc;
```
Comment (clefru) [inline pkgs/top-level/all-packages.nix]: @aszlig: I switched to the form you suggested. I also squashed commit 1+2 together as commit 2 became trivial.

**Comment by fpletz** [comment]
@Mic92 There's already a branch and a hydra jobset. The efforts are going on for a while now and it looks pretty good: https://hydra.nixos.org/jobset/nixos/gcc-7

**Comment by clefru** [comment]
This is security problem. Can we get this merged soon? Then backport to 17.09. Then we can merge the gcc-7 branch into master, and rollback the last commit in master, leaving 17.09 with the backport.

**Comment by vcunat** [comment]
Yes, I agree, do this now and revert later.  The gcc7 branch will probably need some additional work (I see ~1.5k build regressions on Linux platforms ATM), and this PR is relatively cheap rebuild-wise.

Last call!

**Comment by vcunat** [comment]
BTW, this will need nontrivial backporting work for 17.09, and we won't have gcc7 by default there.

**Comment by clefru** [comment]
At the risk of being annoying, this 1.severity: security PR is still open. It puts the kernel in the default state under gcc 7.3, which is retpoline enabled, which is what the kernel devs thought would be a good default despite the performance regressions.

**Comment by edolstra** [comment]
IMHO we should *not* backport this to 17.09. Switching compilers is far outside of the scope of a "stable" branch.

**Comment by vcunat** [comment]
@edolstra: the intention was to backport compiler change _only_ for kernel and its modules, not for all packages (I assume that's what you understood).

**Pull request closed without merging**